packages/ens-utils/src/ethregistrar.ts

```diff
@@ -3,323 +3,370 @@ import {
   MIN_ETH_REGISTRABLE_LABEL_LENGTH,
   ETH_TLD,
   charCount,
-  getDomainLabelFromENSName,
+  buildENSName,
 } from "./ensname";
-import { NFTRef, TokenId, buildNFTRef, buildTokenId } from "./nft";
-import { namehash, labelhash } from "viem/ens";
-import { buildAddress } from "./address";
-import { ChainId, MAINNET } from "./chain";
-import { ContractRef, buildContractRef } from "./contract";
+import { buildContractRef, ContractRef } from "./contract";
 import {
   Duration,
   SECONDS_PER_DAY,
   Timestamp,
   addSeconds,
   buildDuration,
+  buildTimePeriod,
   formatTimestampAsDistanceToNow,
   now,
 } from "./time";
 import {
   Price,
   addPrices,
   approxScalePrice,
+  buildPrice,
   formattedPrice,
-  multiplyPriceByNumber,
   subtractPrices,
 } from "./price";
 import { Currency } from "./currency";
+import {
+  PrimaryRegistrationStatus,
+  RegistrarChargeType,
+  RegistrarUnsupportedNameError,
+  Registration,
+  scaleAnnualPrice,
+  SecondaryRegistrationStatus,
+  RegistrarCharge,
+  RegistrarTemporaryFee,
+  RegistrarBaseFee,
+  RegistrarSpecialNameFee,
+  RegistrationPriceQuote,
+  RegistrarAction,
+  RenewalPriceQuote,
+  OnchainRegistrar,
+  MAINNET_ENS_REGISTRY_WITH_FALLBACK_CONTRACT,
+} from "./registrar";
+import { MAINNET } from "./chain";
+import { buildAddress } from "./address";
 
-export interface Registrar {
-  contract: ContractRef;
-
-  getTokenId(name: ENSName, isWrapped: boolean): TokenId;
-  isClaimable(name: ENSName, isWrapped: boolean): boolean;
-}
+/**
+ * The `EthRegistrar` models the policy implmentations shared by both of
```

Comment:
```suggestion
 * The `EthRegistrar` models the policy implementations shared by both of
```

Reply:
@lightwalker-eth If I can create my own `EthRegistrar` that registers `.eth`, what are the conditions to do so in a valid way?
 
Make this Registrar aligned to `BaseRegistrar`?

For learning purpose, not a high priority to answer!